### PR TITLE
Publish-PSResource Script publishing support

### PR DESCRIPTION
**Pull request by anamnavi**
# PR Summary

Please ignore the branch name. 

This PR will ensure we can publish script resources (locally and remotely) and add tests for local publishing. For publishing to a local path, if it's not a server based path but rather a Drive based path, we also check that the directory path exists.

For testing scripts that have errors, I created a `testFiles` folder under `test` and have added invalid scripts there. Otherwise the script would have been created with New-ScriptFileInfo cmdlet and that cmdlet would have caught the errors instead of letting Publish-PSResource error handle it.

It also adds error handling and bool return for PackNupkg() and PushNupkg() methods.

## PR Checklist

- [ ] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [ ] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [ ] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [ ] None
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [ ] [Documentation needed]()
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [ ] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [ ] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and enumerated concerns in the summary.

### CODE REVIEW

File: src/code/PublishPSResource.cs [hunk 1]
```diff
@@ -61,10 +62,12 @@ public string Path
 
                 if (Directory.Exists(resolvedPath))
                 {
+                    // we point to a folder when publishing a module
                     _path = resolvedPath;
                 }
-                else if (File.Exists(resolvedPath) && resolvedPath.EndsWith(PSDataFileExt, StringComparison.OrdinalIgnoreCase))
+                else if (File.Exists(resolvedPath) && resolvedPath.EndsWith(PSScriptFileExt, StringComparison.OrdinalIgnoreCase))
```
Comment (PaulHigin) [inline src/code/PublishPSResource.cs]: What if an unsupported filepath is provided?  Is an error thrown?  It seems like the code to determine whether the path is directory or supported file path, along with all validation, should be done in the BeginProcess() block.

File: src/code/PublishPSResource.cs [hunk 2]
```diff
@@ -197,31 +200,41 @@ protected override void ProcessRecord()
...
-                    var message = "No version was provided in the script metadata. Script metadata must specify a version, author and description.";
+                    var message = "No version was provided in the script metadata. Script metadata must specify a version, author, description, and Guid.";
                     var ex = new ArgumentException(message);
                     var InvalidScriptMetadata = new ErrorRecord(ex, "InvalidScriptMetadata", ErrorCategory.InvalidData, null);
                     WriteError(InvalidScriptMetadata);
 
                     return;
                 }
-                if (!parsedMetadataHash.ContainsKey("Author") || String.IsNullOrWhiteSpace(parsedMetadataHash["Author"].ToString()))
+                if (!parsedMetadataHash.ContainsKey("author") || String.IsNullOrWhiteSpace(parsedMetadataHash["author"].ToString()))
```
Comment (PaulHigin) [inline src/code/PublishPSResource.cs]: A blank line should always be included after a closing bracket.

File: src/code/PublishPSResource.cs [hunk 3]
```diff
@@ -366,10 +379,10 @@ protected override void ProcessRecord()
                 // pack into a nupkg
                 try
                 {
-                     if (!PackNupkg(outputDir, outputNupkgDir, nuspec))
-                     {
-                        return;
-                     }
+                    if (!PackNupkg(outputDir, outputNupkgDir, nuspec))
```
Comment (PaulHigin) [inline src/code/PublishPSResource.cs]: What does 'false' returned here mean?  Should some kind of error or verbose message be emitted?

File: src/code/PublishPSResource.cs [hunk 4]
```diff
@@ -783,6 +797,138 @@ .PARAMETER Name
             return parsedMetadataHash;
         }
 
+        private Hashtable ParseScriptMetadataIntoHash(string filePath)
```
Comment (PaulHigin) [inline src/code/PublishPSResource.cs]:
```suggestion
        private Hashtable ParseScriptMetadata(string filePath)

```

File: src/code/PublishPSResource.cs [hunk 5]
```diff
@@ -783,6 +797,138 @@ .PARAMETER Name
...
+            var ast = Parser.ParseFile(
+                filePath,
+                out System.Management.Automation.Language.Token[] tokens,
+                out ParseError[] parserErrors);
+
+            if (parserErrors.Length > 0 && !String.Equals(parserErrors[0].ErrorId, "WorkflowNotSupportedInPowerShellCore", StringComparison.OrdinalIgnoreCase))
+            {
+                // TODO: do we want to completely ignore WorkFlowNotSupportedInPowerShellCore error (not even write)
+                // or do we want to write, but not return if it's just that error?
+                foreach (ParseError err in parserErrors)
```
Comment (PaulHigin) [inline src/code/PublishPSResource.cs]: If we want to skip workflow parse errors, then that can be added within the foreach loop.  Also we should return after all parse errors are emitted.

File: src/code/PublishPSResource.cs [hunk 6]
```diff
@@ -783,6 +797,138 @@ .PARAMETER Name
...
+                    var message = String.Format("Could not parse '{0}' as a PowerShell script file due to {1}.", filePath, err.Message);
+                    var ex = new ArgumentException(message);
+                    var psScriptFileParseError = new ErrorRecord(ex, err.ErrorId, ErrorCategory.ParserError, null);
+                    WriteError(psScriptFileParseError);
+                    return parsedPSScriptInfoHashtable;
+
+                }
+
+            }
+            else if (ast != null)
```
Comment (PaulHigin) [inline src/code/PublishPSResource.cs]: ast == null should be an early out.

File: src/code/PublishPSResource.cs [hunk 7]
```diff
@@ -783,6 +797,138 @@ .PARAMETER Name
...
+            {
+                // Get the block/group comment beginning with <#PSScriptInfo
+                List<System.Management.Automation.Language.Token> commentTokens = tokens.Where(a => String.Equals(a.Kind.ToString(), "Comment", StringComparison.OrdinalIgnoreCase)).ToList();
+                string commentPattern = "<#PSScriptInfo";
```
Comment (PaulHigin) [inline src/code/PublishPSResource.cs]: This can be const string.

**Review by PaulHigin — CHANGES_REQUESTED** [review]
Still reviewing...

File: src/code/PublishPSResource.cs (hunk 1 above)
Comment (anamnavi) [reply src/code/PublishPSResource.cs]: added an `else` block to ThrowTerminatingError() if not a valid path for script or module resources in the path format we currently expect

File: src/code/PublishPSResource.cs (hunk 3 above)
Comment (anamnavi) [reply src/code/PublishPSResource.cs]: need to ask Amber

File: src/code/PublishPSResource.cs (hunk 4 above)
Comment (anamnavi) [reply src/code/PublishPSResource.cs]: renamed to TryParseScriptMetadata()

File: src/code/PublishPSResource.cs [hunk 8]
```diff
@@ -190,41 +187,23 @@ protected override void ProcessRecord()
             }
 
             string resourceFilePath;
-            Hashtable parsedMetadataHash = new Hashtable(StringComparer.InvariantCultureIgnoreCase);
+            Hashtable parsedMetadata;
             if (isScript)
             {
+                // this converts from System.IO.DirectoryInfo to string. resourceFilePath needs to be set for later use though
+                // alternatively could use resourceFilePath = _path
```
Comment (anamnavi) [inline src/code/PublishPSResource.cs]: is it faster/more efficient to use the variable (pkgFileOrDir) to set resourceFilePath, or use the private field (_path) to set the resourceFilePath variable?

File: src/code/PublishPSResource.cs (hunk 3 above)
Comment (anamnavi) [reply src/code/PublishPSResource.cs]: I changed the method to return a Error if it fails to pack the package into a .nupkg and to write that error and return when execution returns to EndProcessing().

File: src/code/PublishPSResource.cs (hunk 8 above)
Comment (alerickson) [reply src/code/PublishPSResource.cs]: Hmm, I don't really have any opinions on this one.  I don't see one as being more advantageous than the other.

File: src/code/PublishPSResource.cs [hunk 9]
```diff
@@ -713,74 +697,157 @@ private Hashtable ParseRequiredModules(Hashtable parsedMetadataHash)
             return dependenciesHash;
         }
 
-        private Hashtable ParseScriptMetadata(string filePath)
+        private bool TryParseScriptMetadata(
+            out Hashtable parsedMetadata,
+            string filePath,
+            out ErrorRecord[] errors)
         {
-            // parse .ps1 - example .ps1 metadata:
```
Comment (alerickson) [inline src/code/PublishPSResource.cs:742]: I think we should leave this example metadata in the comment just to have an idea of what a proper script should look like.

File: test/PSGetTestUtils.psm1 [hunk 10]
```diff
@@ -315,33 +315,41 @@ function Get-ScriptResourcePublishedToLocalRepoTestDrive
...
     Get-TestDriveSetUp
 
+    # $publishScriptBase = Join-Path $script:testIndividualResourceFolder $scriptName
+    # $null = New-Item -Path $publishScriptBase -ItemType Directory -Force
+
     $scriptFilePath = Join-Path -Path $script:testIndividualResourceFolder -ChildPath "$scriptName.ps1"
     $null = New-Item -Path $scriptFilePath -ItemType File -Force
 
-    $version = "1.0.0"
+    # $version = "1.0.0"
```
Comment (alerickson) [inline test/PSGetTestUtils.psm1]: This looks like it might not be needed anymore

File: test/PSGetTestUtils.psm1 [hunk 11]
```diff
@@ -315,33 +315,41 @@ function Get-ScriptResourcePublishedToLocalRepoTestDrive
...
+
+        [string]
+        $scriptRepoName,
+
+        [string]
+        $scriptVersion
     )
     Get-TestDriveSetUp
 
+    # $publishScriptBase = Join-Path $script:testIndividualResourceFolder $scriptName
```
Comment (alerickson) [inline test/PSGetTestUtils.psm1]:
```suggestion
```

File: test/PSGetTestUtils.psm1 [hunk 12]
```diff
@@ -315,33 +315,41 @@ function Get-ScriptResourcePublishedToLocalRepoTestDrive
...
+    # $null = New-Item -Path $publishScriptBase -ItemType Directory -Force
```
Comment (alerickson) [inline test/PSGetTestUtils.psm1]:
```suggestion
```

File: test/PSGetTestUtils.psm1 [hunk 13]
```diff
@@ -315,33 +315,41 @@ function Get-ScriptResourcePublishedToLocalRepoTestDrive
...
+
```
Comment (alerickson) [inline test/PSGetTestUtils.psm1]:
```suggestion
```

File: test/PSGetTestUtils.psm1 (hunk 10 above)
Comment (alerickson) [inline test/PSGetTestUtils.psm1]:
```suggestion
```

File: test/PSGetTestUtils.psm1 [hunk 14]
```diff
@@ -525,9 +531,15 @@ function Create-PSScriptMetadata
 
 .RELEASENOTES
 $($ReleaseNotes -join "`r`n")
-
 .PRIVATEDATA$(if ($PrivateData) {" $PrivateData"})
```
Comment (alerickson) [inline test/PSGetTestUtils.psm1:530]:
```suggestion

.PRIVATEDATA$(if ($PrivateData) {" $PrivateData"})
```

File: test/PSGetTestUtils.psm1 [hunk 15]
```diff
@@ -525,9 +531,15 @@ function Create-PSScriptMetadata
...
 
+#>
+
+<#
+
```
Comment (alerickson) [inline test/PSGetTestUtils.psm1:535]:
```suggestion
```

File: test/PSGetTestUtils.psm1 [hunk 16]
```diff
@@ -525,9 +531,15 @@ function Create-PSScriptMetadata
...
+.DESCRIPTION
+$(if ($Description) {" $Description"})
+
```
Comment (alerickson) [inline test/PSGetTestUtils.psm1:538]:
```suggestion
```

File: test/UninstallPSResource.Tests.ps1 [hunk 17]
```diff
@@ -287,4 +287,4 @@ Describe 'Test Uninstall-PSResource for Modules' {
         $pkg.Name | Should -Be $testModuleName
         $pkg.Path.ToString().Contains("Documents") | Should -Be $true
     }
-}
+}
```
Comment (alerickson) [inline test/UninstallPSResource.Tests.ps1]:
```suggestion
}
```

**Review by alerickson — APPROVED** [review]
Looks good, just left some suggestions for removing comments and whitespace, please take them at your discretion!

File: test/PSGetTestUtils.psm1 (hunk 15 above)
Comment (anamnavi) [reply test/PSGetTestUtils.psm1:535]: this is weird .ps1 file script syntax...for the `ast.GetHelpContent()` method to pick up `.DESCRIPTION` that comment block (i.e `<#`) has to be preceeded by a space, and has to be followed by a space before the `.DESCRIPTION` key, and then again before the closing comment block (i.e `#>`) it needs another space. We should have a blogpost or documentation somewhere about these weird format peculiarities .ps1 files have

File: test/PSGetTestUtils.psm1 (hunk 16 above)
Comment (anamnavi) [reply test/PSGetTestUtils.psm1:538]: this is weird .ps1 file script syntax...for the ast.GetHelpContent() method to pick up .DESCRIPTION that comment block (i.e <#) has to be preceeded by a space, and has to be followed by a space before the .DESCRIPTION key, and then again before the closing comment block (i.e #>) it needs another space. We should have a blogpost or documentation somewhere about these weird format peculiarities .ps1 files have

File: src/code/PublishPSResource.cs (hunk 8 above)
Comment (anamnavi) [reply src/code/PublishPSResource.cs]: I'll keep this as is

**Pull request merged**